### DISCUSSION TITLE
PR notification link is useless on push force

### DISCUSSION THREAD

**Discussion by xmo-odoo** [community/community#31738] [Pull Requests]
edit: dupe of #5373 which I'd apparently forgotten about, but it doesn't seem like it's possible to close or dupe discussion threads.

If a PR is pushed to, the notification's link shows the diff betweeb the pushed-from and pushed-to commits.

This is OK when adding new commits to a PR, but it's useless any time the PR is reset / force-pushed e.g.

- the PR was rebased to resync with upstream (because "linear history" was required for instance)
- the PR was edited to clean up its history when prepping for merge (e.g. squashed everything)
- fixes were integrated via `rebase -i` (e.g. fixup)

In all those cases, because the new head is not a descendent of the old head github links to an empty diff view, *which it knows will be empty*, with

> We went looking everywhere, but couldn’t find those commits.
>
> Sometimes commits can disappear after a force-push. Head back to the latest changes here.

Github actually provides a *sometimes* useful textual diff in the

> @user force-pushed the thing branch from a to b some time ago

the "force pushed" words link to the `/compare` view between the `a` and `b`. This is completely useless when rebasing on upstream (as all the changes *of upstream* will be visible), but it at least has some odds of being informative.

**Comment by benoittgt** [comment]
It should definitly need to display the files changed but with a toast message "force pushed"

**Comment by AndiDog** [comment]
It seems that the annoying **We went looking everywhere, but couldn’t find those commits.** only happens in one diff view of the GitHub UI, but not another, so the commits aren't lost.

Here's a **real example why GitHub _could_ actually show the diff**. It's from [this PR](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5675), but I'm not sure how long after merge you'll still be able to see the same. The author made some rebases, and GitHub shows the from-to commit hashes:

![Screenshot 2025-10-16 at 18 57 26](https://github.com/user-attachments/assets/31273290-6090-4b3c-b00b-757cc8037c54)

The "compare" link is of the form `https://github.com/kubernetes-sigs/cluster-api-provider-aws/compare/0c130b1ee3ab766d93e83ae8fee3de8792c75ce9..868be8c2ef7fbbe36865e64278b3739b49a47c01` **and works**.

Now as a reviewer, I want to see what changed since my last review, so I switch to "Files changed > All commits dropdown > Changes since your last review" (btw I'm on the new PR experience already). And I get the dreaded "We went looking everywhere, but couldn’t find those commits.". The link for this view is `https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/5675/files/10bc249b497970f70980b3266d85359c5ccfe119..090a6d9afed3cee6deffc21d4d6a17aa86efac44` so this **doesn't work**.

I'm not giving up though. Let's fill the from-to commits from the "Changes since your last review" link into the `/compare/<from>..<to>` URL: `https://github.com/kubernetes-sigs/cluster-api-provider-aws/compare/10bc249b497970f70980b3266d85359c5ccfe119..090a6d9afed3cee6deffc21d4d6a17aa86efac44`. And tada, **this works**. The error message therefore is a just a lie – GitHub has the commits. The downside of this diff, of course, is that you see all changes from the rebase, not only the ones pertaining to the PR, so I get why GitHub probably doesn't want to display that potentially-large diff.

A diff-of-diffs view would be such a win for all reviewers around the globe. Phabricator Differential, which has been dead for a long time, had exactly that feature, and it was really great to quickly review stuff despite rebase. So for now, as a reviewer on GitHub, I have to use this trick to change URLs, and then jump to the files changed in the PR, which often gives me a good picture what the author has changed vs. what came in through the rebase.

Here's an example how it _could_ look like: [example review in Phabricator Differential](https://reviews.freebsd.org/D53109). Go to "Revision Contents > History" and under "Base", you'll see the parent commit of each revision of the review. The review got rebased once. Click the radio buttons on the right to see the difference between "Diff 1" and "Diff 2" (= the last revision done to the review). And you'll see it was an empty rebase. You don't see all the stuff that came in by rebasing. It's a real diff-of-diffs. The source code must be somewhere in the still-existing [Phabricator repo](https://github.com/phacility/phabricator/tree/master/src/applications/differential/engine). I'm really longing to have that functionality back...

(On the same note: yeah there's `git range-diff` but the important part is how to display that nicely)
